### ISSUE TITLE
D3 scatter plot - test #8 failing

### ISSUE THREAD

**Issue by tom-p-uk** [freeCodeCamp/testable-projects-fcc#70]
#### Issue Description

Built this graph today but it's failing one test when it shouldn't be.

It's failing test number 8 of 15. I checked out the criteria for the test by visiting the test JS file [here](https://gitcdn.link/cdn/freeCodeCamp/testable-projects-fcc/1de9ee9c2350f4e1963bc061bba7ffc56221e4cc/build/bundle.js) and then did a little bit of debugging.

It turns out that my scatter plot fails the test because the sort function in the test page sorts my array of dots in descending order rather than ascending, which means it fails at the very first assertion. If you look at the code and the chart itself, though, it's clearly all in working order.

#### Your Code / Link to Your Pen

[Codepen link](http://codepen.io/tom-p-uk/full/jBbMqq/)

Code for the appropriate failing test:
```
it('8. The data-yvalue and its corresponding dot should align with the corresponding point/value on the y-axis.', function () {
	                var dotsCollection = document.getElementsByClassName('dot');
	                //convert to array
	                var dots = [].slice.call(dotsCollection);
	                FCC_Global.assert.isAbove(dots.length, 0, 'there are no elements with the class of "dot" ');
	                //sort the dots based on yvalue in ascending order
	                var sortedDots = dots.sort(function (a, b) {
	                    return new Date(a.getAttribute("data-yvalue")) - new Date(b.getAttribute("data-yvalue"));
	                });

	                //check to see if the y locations of the new sorted array are in ascending order
	                for (var i = 0; i < sortedDots.length - 1; ++i) {
	                    FCC_Global.assert.isAtMost(+sortedDots[i].cy.baseVal.value, +sortedDots[i + 1].cy.baseVal.value, "y values don't line up with y locations ");
	                }
	            });

```

**Comment by no-stack-dub-sack** [comment]
@paycoguy @Christian-Paul ☝️

**Comment by paycoguy** [comment]
@tom-p-uk Thanks Tom for building these D3 projects!  I believe we can validate the scatter plot whether the axis values are ascending or descending.  It'll just make the code a bit more complicated.

**Comment by tchaffee** [comment]
@paycoguy @Christian-Paul @QuincyLarson I need some direction for this failing test. @no-stack-dub-sack you might want to chime in?

The test does not do what it says it does. The test says "The data-yvalue and its corresponding dot should align with the corresponding point/value on the y-axis." But there is no code in that test that checks alignment with the y-axis. What it does, as @tom-p-uk points out, is to check if the `cy` values (vertical position) of the dots are in ascending order *after sorting the dots by `data-yvalue`*. In plain English, it checks if dots with a *smaller* data-yvalue (faster time) appear *above* dots with larger `data-yvalue`.

I looked into seeing if I could check if the dots align correctly with the y-axis labels and it looks pretty complicated. Important also that d3 does that work for you, so I'm not sure how much value that test has.

In the meantime, I think the best way I can do is to fix the existing test wording so it better describes what it is testing. It would clarify what's being testing, along with making sure the values are ascending:

"The vertical position of each dot must be related to the data-yvalue attribute of the dot, and dots with a faster time (smaller data-yvalue) should be above dots with a longer time (larger data-yvalue)"

This test, along with test 12, "I can see that the range of the y-axis labels are within the range of the actual y-axis data" should be enough to ensure the dots are aligned with the y-axis, considering that d3 is the one doing that work for the camper.

Thoughts please?

**Comment by tchaffee** [comment]
@tbushman I forgot you are working on some d3 issues yourself. If you have any ideas on the above, your comments are of course welcome.

**Comment by tbushman** [comment]
@tchaffee I was looking into splitting up the .dot 'd' attribute to compare moveto x,y to position on the axis path, because as you say, the test doesn't seem to check for alignment, only the sorted order. I'll try and come up with something to share in a little while. Thanks.

**Comment by tchaffee** [comment]
@tbushman that's excellent because I was stuck. You're thinking about this more creatively than I am and I'm looking forward to seeing what you come up with.

**Comment by paycoguy** [comment]
> I looked into seeing if I could check if the dots align correctly with the y-axis labels and it looks pretty complicated.

We tried to do this as well and couldn't find a good way to do it.  

I believe I tried making this test work whether the axis was ascending or descending, but the resulting code became really long.

> "The vertical position of each dot must be related to the data-yvalue attribute of the dot, and dots with a faster time (smaller data-yvalue) should be above dots with a longer time (larger data-yvalue)"

This sounds alright with me.

**Comment by tbushman** [comment]
Reopening due to revert.

**Comment by tbushman** [comment]
Codepen appears to handle the 0.5px offset on axes automatically generated by D3 differently than in the regular window. I closed #137 because while it worked in the rawgit bundle.js imported into Codepen, the local tests had a few misalignments counted that were off by 0.5px. The final PR coming soon should compare feature coordinates to axis tick coordinates within a small range (even + / - 1 would do it) instead of exact..
